nucypher/characters/base.py

```diff
@@ -355,53 +365,68 @@ def verify_from(self,
+            sender_verifying_key = stranger.stamp.as_umbral_pubkey()
+            if message_kit.sender_verifying_key:
+                if not message_kit.sender_verifying_key == sender_verifying_key:
+                    raise ValueError("This MessageKit doesn't appear to have come from {}".format(stranger))
+
+        #
+        # Decrypt
+        #
+
+        signature_from_kit = None
         if decrypt:
 
             # We are decrypting the message; let's do that first and see what the sig header says.
-            cleartext_with_sig_header = self.decrypt(message_kit=message_kit,
-                                                     label=label)
+            cleartext_with_sig_header = self.decrypt(message_kit=message_kit, label=label)
             sig_header, cleartext = default_constant_splitter(cleartext_with_sig_header, return_remainder=True)
+
             if sig_header == SIGNATURE_IS_ON_CIPHERTEXT:
-                # THe ciphertext is what is signed - note that for later.
+                # The ciphertext is what is signed - note that for later.
                 message = message_kit.ciphertext
                 if not signature:
                     raise ValueError("Can't check a signature on the ciphertext if don't provide one.")
+
             elif sig_header == SIGNATURE_TO_FOLLOW:
                 # The signature follows in this cleartext - split it off.
-                signature_from_kit, cleartext = signature_splitter(cleartext,
-                                                                   return_remainder=True)
+                signature_from_kit, cleartext = signature_splitter(cleartext, return_remainder=True)
                 message = cleartext
+
         else:
             # Not decrypting - the message is the object passed in as a message kit.  Cast it.
             message = bytes(message_kit)
             cleartext = NO_DECRYPTION_PERFORMED
 
+        #
+        # Verify Signature
+        #
+
         if signature and signature_from_kit:
             if signature != signature_from_kit:
                 raise ValueError(
                     "The MessageKit has a Signature, but it's not the same one you provided.  Something's up.")
 
         signature_to_use = signature or signature_from_kit
-
         if signature_to_use:
-            is_valid = signature_to_use.verify(message, sender_pubkey_sig)
+            is_valid = signature_to_use.verify(message, sender_verifying_key)  # FIXME: Message is undefined here
             if not is_valid:
-                raise stranger.InvalidSignature(
-                    "Signature for message isn't valid: {}".format(signature_to_use))
+                raise InvalidSignature("Signature for message isn't valid: {}".format(signature_to_use))
         else:
-            raise self.InvalidSignature("No signature provided -- signature presumed invalid.")
+            raise InvalidSignature("No signature provided -- signature presumed invalid.")
```

Comment:
Why are we changing here the type of `InvalidSignature`? Previous one was our own, and now it's the one from the cryptography, even if the `Signature` is ours.

Reply:
See related PR #1040